# Artificial Neural Network - Vj

## Part 1 - Data Preprocessing

### Importing the libraries

In [ ]:
import numpy as np
import tensorflow as tf
import pandas as pd

Check Version of Tensorflow

In [ ]:
tf.__version__

'2.4.1'

### Importing dataset


In [ ]:
dataset = pd.read_csv('Churn_Modelling.csv')
# Exclude first 3 columns and the last column, because they are not useful.
# Index starts at 0.
x = dataset.iloc[:, 3:-1].values

# Just get the last column of dataset for the dependent variable.
y = dataset.iloc[:, -1].values

FileNotFoundError: ignored

In [ ]:
print(x)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [ ]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
# Get all the rwos for the 2 column and encode them.
x[:, 2] = le.fit_transform(x[:, 2])

In [ ]:
print(x)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column

In [ ]:
# We do one hot coding for geographical places because they have no relation among themselves like gender did.
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough')
x = np.array(ct.fit_transform(x))

In [ ]:
print (x)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 1)

### Feature Scaling

In [ ]:
# pply feature scaling to all the features of both the training and the test set.
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
# Scaling fitted only to training set to avoid information leakage.
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initializing the ANN

In [ ]:
# Create a var to represtn the ANN as an instance of the sequential class that initializes our ANN.
ann = tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [ ]:
# add() method belongs to the sequential class.
# add a fully connected layer (hidden), which will be a new object for the dense class.
# Argument for dense() - 
# 'units' - There are called hyperparameters and are the hidden neurons.
# 'activation' - activation function in the hidden layer must be a rectifier activation function.
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) # 'relu' - codename for rectifier activation function.

### Adding the second hidden layer

In [ ]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu')) 

### Adding the output layer

In [ ]:
# As the output layer is binary we only need one neuron to match with the output layer, so 'units' = 1
# We need a sigmoid activation function, because it will give the predictions, as well as the probabilities.
# If we were predicting a categorical variable, activation = 'sigmax'
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid')) 

## Part 3 - Training the ANN

### Compiling the ANN

In [ ]:
# Arguments - 
# 'optimizer' - We choose 'adam' optimizer as it can perform stochastic gradient descent. 
#  The stochastic gradient descent will optimize the weights to minimize the loss between the actual and the predicted values.
# 'loss' - When we are doing a binary classification when we are predicting a binary variable use 'binary_crossentropy' and 
#  if we had categorical variable to predict we would have used 'categorical_crossentropy'
# 'metrics' - 
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

### Training the ANN on the Training set

In [ ]:
# Batch learning, default batch size to be compared to the actual results = 32
ann.fit(X_train, y_train, batch_size=32, epochs=100)
# As we can see the accuracy converges around 0.86 around epoch = 20, so we didn't need 100 epochs.

Epoch 1/100
250/250 [==============================] - 0s 1ms/step - loss: 0.5126 - accuracy: 0.7956
Epoch 2/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4506 - accuracy: 0.7972
Epoch 3/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4380 - accuracy: 0.7972
Epoch 4/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4331 - accuracy: 0.7972
Epoch 5/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4301 - accuracy: 0.7972
Epoch 6/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4281 - accuracy: 0.7972
Epoch 7/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4262 - accuracy: 0.7972
Epoch 8/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4248 - accuracy: 0.7972
Epoch 9/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4230 - accuracy: 0.7972
Epoch 10/100
250/250 [==============================] - 0s 1ms/step - loss: 0.4211 - accura

## Part 4 - Making the predictions and evaluating the model

### Predicting the result of a single observation

Using this ANN model to predict if the customer with the following informations will leave the bank: 

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: \$ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: \$ 50000

So, should we say goodbye to that customer?

**Solution**

In [ ]:
# As we used a sigmoid function, we get the probability of this person leaving the bank.
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])))

[[0.03832316]]


Therefore, our ANN model predicts that this customer stays in the bank!

**Important note 1:** Notice that the values of the features were all input in a double pair of square brackets. That's because the "predict" method always expects a 2D array as the format of its inputs. And putting our values into a double pair of square brackets makes the input exactly a 2D array.

**Important note 2:** Notice also that the "France" country was not input as a string in the last column but as "1, 0, 0" in the first three columns. That's because of course the predict method expects the one-hot-encoded values of the state, and as we see in the first row of the matrix of features X, "France" was encoded as "1, 0, 0". And be careful to include these values in the first three columns, because the dummy variables are always created in the first columns.

### Predicting the Test set results

In [ ]:
# These are predicted probabilities
y_predicted = ann.predict(X_test)
# Convert Predicted probabilities into binary outcome
y_predicted = (y_predicted > 0.5)
print(np.concatenate((y_predicted.reshape(len(y_predicted),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 0]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_predicted)
print(cm)
accuracy_score(y_test, y_predicted)

[[1541   44]
 [ 230  185]]


0.863

Therefore, Accuracy = 86%

Total correct predictions that customers stay in bank = 1541

Total correct predictions that customers leave the bank = 185

Total incorrect predictions that customers stay in the bank = 230

Total incorrect predictions that customers leave in the bank = 44